### DISCUSSION TITLE
Return dictionary from Rust to Python

### DISCUSSION THREAD

**Discussion by cassc** [PyO3/pyo3#3078] [Questions] [answered]
In `#[pymethods]`, is it possible to return a Python dict? It looks like creating [PyDict](https://pyo3.rs/main/doc/pyo3/types/struct.pydict) would require access to the global GIL. However, it's not clear how to get the GIL from a function in `#[pymthods]` or `#[pyclass]` other than using the deprecated `Python::acquire_gil()`

**Comment by adamreichold** [comment, marked as answer]
> However, it's not clear how to get the GIL from a function in #[pymthods] or #[pyclass] other than using the deprecated Python::acquire_gil()

There are two answers to this: The replacement for `Python::acquire_gil` is [`Python::with_gil`](https://docs.rs/pyo3/latest/pyo3/marker/struct.Python.html#method.with_gil) which serves the same purpose using a slightly different API (that is easier for us to implement soundly).

However, for `#[pyfunction]`s and `#[pymethod]`s, you can just add a parameter `py: Python<'_>` to get access to the GIL (which is already acquired when such a method is called). Put differently, `Python::with_gil` is only necessary outside of such Python-visible functions and methods, for example to acquire the GIL on a background thread.

**Reply by adamreichold** [reply]
Just to add an example:

```rust
#[pyfunction]
fn make_dict<'py>(py: Python<'py>, value: i32) -> PyResult<&'py PyDict> {
    let dict = PyDict::new(py);

    dict.set_item("value", value)?;

    Ok(dict)
}
```

The documentation also contains some hints on how to obtain a `Python` token under various circumstances, c.f. https://docs.rs/pyo3/latest/pyo3/marker/struct.Python.html#obtaining-a-python-token

**Comment by cassc** [comment]
This is perfect, thanks so much!
